### PR TITLE
 Add a 'run' link to all examples in the documenation

### PR DESCRIPTION
**Pull request by Bzero**
Should resolve issue #485.

### CODE REVIEW

File: book/build.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    path_in = SCRIPT_DIR.parent / "examples" / f"{filename}.nbt"
+    path_out = SCRIPT_DIR / "src" / f"example-{filename}.md"
+
+    code = []
+    with open(path_in, "r") as fin:
+        for line in fin:
+            if not strip_assets or "assert_eq" not in line:
+                code.append(line)
+
+    url = f"https://numbat.dev/?q={urllib.parse.quote_plus(''.join(code))}"
```
Comment (sharkdp) [inline book/build.py:19]: Was about to comment that this doesn't work on Python 3.10 locally for me, but you already fixed that :+1:

File: book/build.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+    with open(path_out, "w") as fout:
+        fout.write("<!-- This file is autogenerated! Do not modify it -->\n")
+        fout.write("\n")
+        fout.write(f'# {title} <a href="{url}"><i class="fa fa-play"></i></a>\n')
```
Comment (sharkdp) [inline book/build.py:24]: Nice idea with the icon! Layout-wise, I think I would prefer something a bit more conservative like:
```suggestion
        fout.write(f"# {title}\n")
        fout.write(f'<a href="{url}"><i class="fa fa-play"></i> Run this example</a>\n')
```

File: book/build.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+generate_example("pipe_flow_rate", "Flow rate in a pipe", True)
+generate_example("population_growth", "Population growth", True)
+generate_example("recipe", "Recipe", True)
+generate_example("voyager", "Voyager", True)
+generate_example("xkcd_687", "XKCD 687", True)
+generate_example("xkcd_2585", "XKCD 2585", True)
+generate_example("xkcd_2812", "XKCD 2812", True)
+generate_example("workhours", "Workhours", True)
+
+generate_example("numbat_syntax", "Syntax overview", False)
```
Comment (sharkdp) [inline book/build.py:48]: Running this syntax overview (all at once) does not make a lot of sense. Can we skip the "Run" link for this page? We could add a fourth (defaulted) argument to `generate_example`

File: book/build.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
+import subprocess
+from pathlib import Path
+import urllib.parse
+
+
+SCRIPT_DIR = Path(__file__).parent
+
+
+def generate_example(filename, title, strip_assets):
```
Comment (sharkdp) [inline book/build.py:9]:
```suggestion
def generate_example(filename, title, strip_asserts):
```

**Review by sharkdp — CHANGES_REQUESTED** [review]
Thank you very much — this is excellent!

File: book/build.py (hunk 1 above)
Comment (Bzero) [reply book/build.py:21]: Yes, I forgot that this only works with newer versions.

File: book/build.py (hunk 3 above)
Comment (Bzero) [reply book/build.py]: Agreed, this is a bit much of a wall of code. I added a parameter to turn the link insertion off.

File: book/build.py (hunk 2 above)
Comment (Bzero) [reply book/build.py]: Sounds good to me!

File: book/build.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+generate_example("musical_note_frequency", "Musical note frequency")
+generate_example("paper_size", "Paper sizes")
+generate_example("pipe_flow_rate", "Flow rate in a pipe")
+generate_example("population_growth", "Population growth")
+generate_example("recipe", "Recipe")
+generate_example("voyager", "Voyager")
+generate_example("xkcd_687", "XKCD 687")
+generate_example("xkcd_2585", "XKCD 2585")
+generate_example("xkcd_2812", "XKCD 2812")
+generate_example("workhours", "Workhours")
```
Comment (Bzero) [inline book/build.py:50]: The "workhours" example was missing in 'build.sh', I would think this was not intentional, right?

Comment (sharkdp) [reply book/build.py]: It was. Notice that there are a lot more examples in `examples/` that are not part of the documentation. I only hand-selected a few of them. I removed this line again.

**Review by sharkdp — APPROVED** [review]
Thank you for the updates. Your changes are live: https://numbat.dev/doc/example-voyager.html

File: book/build.py (hunk 5 above)
Comment (Bzero) [reply book/build.py]: Yes, I noticed that, but it is the only example that is not present in `build.sh`/`build.py` but has an [associated file](https://github.com/sharkdp/numbat/blob/4709bc3d9b6a546a0d90b77578c54bd41faf5167/book/src/example-workhours.md) in `/book/src/`. Maybe that file should be removed too in this case?

Comment (sharkdp) [reply book/build.py]: Oh, right.  I guess it was part of the documentation at some point but then I removed it. I'll remove that file. Thank you.

**Pull request merged**